File: 1282-group-the-people-given-the-group-size-they-belong-to/1282-group-the-people-given-the-group-size-they-belong-to.cpp
class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
        vector<vector<int>> result;
        unordered_map<int, vector<int>> map;
        
        for (int i = 0; i < groupSizes.size(); i++) {
            int size = groupSizes[i];
            if (map.find(size) != map.end()) {
                map[size].push_back(i);
            } else {
                map[size] = {i};
            }
        }
        
        for (auto& kv : map) {
            vector<int>& group = kv.second;
            if (kv.first == group.size()) {
                result.push_back(group);
            } else {
                for (int i = 0; i < group.size(); i += kv.first) {
                    result.push_back(vector<int>(group.begin() + i, group.begin() + i + kv.first));
                }
            }
        }
        return result;
    }
};

File: 1282-group-the-people-given-the-group-size-they-belong-to/1282-group-the-people-given-the-group-size-they-belong-to.py
class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        result = []
        groups = {}
        
        for i in range(len(groupSizes)):
            size = groupSizes[i]
            if size in groups:
                groups[size].append(i)
            else:
                groups[size] = [i]
        
        for key, group in groups.items():
            if key == len(group):
                result.append(group)
            else:
                result.extend([group[i:i + key] for i in range(0, len(group), key)])
        
        return result
